File: src/rules/terrain.ts
import { z } from "zod";
import type { GameState } from "../schema/state.js";
import type { Entity } from "../schema/entity.js";


/**
 * GROUND THAT MATTERS.
 *
 * Every fight in this engine happened on the same floor. A zone was `{id, name, adjacent}`
 * — a label and a graph edge — so "the moot stone" and "the burning roof" played
 * identically, and the only variable in any fight was how many hit points each side had.
 * That is why combat felt the same every time: it WAS the same every time.
 *
 * Terrain is the cheapest variety there is, because it changes decisions rather than
 * numbers. Cover makes a zone worth crossing to; a hazard makes one worth pushing somebody
 * into; high ground makes the first move matter. None of it needs new actions — it reprices
 * the ones already there.
 *
 * Deliberately small: five traits, each one sentence, each with a mechanical hook the
 * resolver already understands. A trait a player cannot see the effect of is set dressing.
 */

export const TerrainTrait = z.enum([
  /** Something to get behind. Harder to hit anyone standing here. */
  "cover",
  /** Above everything. Easier to hit anyone who is not. */
  "high",
  /** Mud, rubble, water. Crossing out of here costs your whole move. */
  "broken",
  /** Fire, a drop, machinery. Being here hurts, and being pushed here hurts more. */
  "hazard",
  /** Shadow, smoke, steam. Easier to hide, harder for anyone to pick you out. */
  "dim",
]);
export type TerrainTrait = z.infer<typeof TerrainTrait>;

export interface TerrainEffect {
  label: string;
  /** One line the client shows on the zone, and the narrator is told about. */
  blurb: string;
}

export const TERRAIN: Record<TerrainTrait, TerrainEffect> = {
  cover:  { label: "cover",      blurb: "something solid to put between you and them" },
  high:   { label: "high ground", blurb: "above the rest of it, with the reach that buys" },
  broken: { label: "broken ground", blurb: "rubble and standing water; leaving costs everything you have" },
  hazard: { label: "hazard",     blurb: "you do not want to be standing here, and neither do they" },
  dim:    { label: "dim",        blurb: "smoke and shadow enough to lose a shape in" },
};

/** Damage taken for ending a turn in a hazard. Small — this is pressure, not a trap. */
export const HAZARD_DAMAGE = 2;

export function traitsOf(s: GameState, locationId: string, zoneId: string | null): TerrainTrait[] {
  if (!zoneId) return [];
  const zone = s.locations[locationId]?.zones.find((z) => z.id === zoneId);
  return zone?.terrain ?? [];
}

/**
 * How the ground changes an attack.
 *
 * Defender's cover and attacker's height, as itemised modifiers so they show up on the
 * roll card with their reasons — which is the only way a player learns that crossing to
 * the rubble was worth a turn.
 */
export function attackModifiers(
  s: GameState,
  attacker: Entity,
  defender: Entity,
): Array<{ label: string; value: number }> {
  const out: Array<{ label: string; value: number }> = [];
  const atk = traitsOf(s, attacker.location_id, attacker.zone_id);
  const def = traitsOf(s, defender.location_id, defender.zone_id);

  if (def.includes("cover")) {
    out.push({ label: "their cover", value: -2 });
  }
  if (def.includes("dim")) {
    out.push({ label: "smoke and shadow", value: -2 });
  }
  // High ground only counts when you are above them, not when everybody is up there.
  if (atk.includes("high") && !def.includes("high")) {
    out.push({ label: "high ground", value: 2 });
  }
  return out;
}

/** A line for the narrator and the client about where somebody is standing. */
export function describeZone(s: GameState, locationId: string, zoneId: string | null): string {
  const traits = traitsOf(s, locationId, zoneId);
  if (traits.length === 0) return "";
  return traits.map((t) => TERRAIN[t].blurb).join("; ");
}

File: _terrain.py
# -*- coding: utf-8 -*-
import io

def load(p):
    raw = io.open(p, encoding="utf-8", newline="").read()
    return ("\r\n" in raw), raw.replace("\r\n", "\n")

def save(p, s, crlf):
    if crlf: s = s.replace("\n", "\r\n")
    io.open(p, "w", encoding="utf-8", newline="").write(s)

def rep(s, old, new, label):
    assert old in s, "anchor: " + label
    return s.replace(old, new, 1)

# ── terrain.ts returns roll PARTS, which is what the roll card already itemises.
p = "src/rules/terrain.ts"
crlf, s = load(p)
s = rep(s,
    'import type { Modifier } from "./modifiers.js";',
    '',
    "drop modifier import")
s = rep(s,
    """export function attackModifiers(
  s: GameState,
  attacker: Entity,
  defender: Entity,
): Modifier[] {
  const out: Modifier[] = [];""",
    """export function attackModifiers(
  s: GameState,
  attacker: Entity,
  defender: Entity,
): Array<{ label: string; value: number }> {
  const out: Array<{ label: string; value: number }> = [];""",
    "signature")
s = s.replace(
    '    out.push({ source: "terrain", reason: `${defender.name} is behind cover`, dc_delta: 0, mod: -2 });',
    '    out.push({ label: "their cover", value: -2 });', 1)
s = s.replace(
    '    out.push({ source: "terrain", reason: `${defender.name} is hard to pick out`, dc_delta: 0, mod: -2 });',
    '    out.push({ label: "smoke and shadow", value: -2 });', 1)
s = s.replace(
    '    out.push({ source: "terrain", reason: "attacking from above", dc_delta: 0, mod: 2 });',
    '    out.push({ label: "high ground", value: 2 });', 1)
save(p, s, crlf)

# ── zones carry terrain
p = "src/schema/location.ts"
crlf, s = load(p)
s = rep(s,
    """export const Zone = z.object({
  id: z.string(),
  name: z.string(),
  adjacent: z.array(z.string()).default([]),
});""",
    """export const Zone = z.object({
  id: z.string(),
  name: z.string(),
  adjacent: z.array(z.string()).default([]),
  /**
   * What the ground does. See rules/terrain.ts.
   *
   * Empty means flat and featureless, which is what every zone used to be — and is why
   * every fight played the same. Terrain is the cheapest variety available, because it
   * changes decisions rather than numbers.
   */
  terrain: z.array(TerrainTrait).default([]),
});""",
    "zone")
s = rep(s, 'import { z } from "zod";',
        'import { z } from "zod";\nimport { TerrainTrait } from "../rules/terrain.js";',
        "import")
save(p, s, crlf)

# ── the attack roll knows about the ground
p = "src/engine/combatActions.ts"
crlf, s = load(p)
s = rep(s,
    """  const atk = rollD20(rng, {
    purpose: "attack", mods: toHit, target: target.ac, isAttack: true, advantage, lean,
    parts: [{ label: ability, value: abilityBonus }, { label: "proficiency", value: attacker.proficiency_bonus }],
  });""",
    """  // The ground, itemised so it reaches the roll card with its reason. A player only
  // learns that crossing to the rubble was worth a turn if the card says so.
  const ground = attackModifiers(s, attacker, target);
  const groundTotal = ground.reduce((n, g) => n + g.value, 0);

  const atk = rollD20(rng, {
    purpose: "attack", mods: toHit + groundTotal, target: target.ac, isAttack: true, advantage, lean,
    parts: [
      { label: ability, value: abilityBonus },
      { label: "proficiency", value: attacker.proficiency_bonus },
      ...ground,
    ],
  });""",
    "roll")
s = rep(s, 'import { conditionFlags } from "../rules/conditions.js";',
        'import { conditionFlags } from "../rules/conditions.js";\nimport { attackModifiers } from "../rules/terrain.js";',
        "import")
save(p, s, crlf)
print("ok")
